evidence.go

```diff
@@ -0,0 +1,279 @@
+		return nil, fmt.Errorf("missing platform claims in evidence")
+	}
+
+	if e.CcaRealmClaims == nil {
+		return nil, fmt.Errorf("missing realm claims in evidence")
+	}
+
+	if err := e.signClaim(PlatformClaim, pSigner); err != nil {
+		return nil, err
+	}
+
+	if err := e.signClaim(RealmClaim, rSigner); err != nil {
+		return nil, err
+	}
+
+	// We do now have CcaPlatform and Realm Token setup correctly.
+	buf, err := em.Marshal(e.message)
+	if err != nil {
+		return nil, fmt.Errorf("cbor encoding of CCA token failed: %w", err)
+	}
+
+	return buf, nil
+}
+
+func (e *CcaEvidence) signClaim(cl claimType, signer cose.Signer) error {
+	var err error
+
+	message := cose.NewSign1Message()
+	if cl == PlatformClaim {
+		message.Payload, err = e.CcaPlatformClaims.ToCBOR()
+		if err != nil {
+			return err
+		}
+	} else if cl == RealmClaim {
+		message.Payload, err = e.CcaRealmClaims.ToCBOR()
+		if err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("invalid claimType %d: for signing", cl)
+	}
+
+	alg := signer.Algorithm()
+
+	if strings.Contains(alg.String(), "unknown algorithm value") {
+		return errors.New("signer has no algorithm")
+	}
+
+	message.Headers.Protected.SetAlgorithm(alg)
+
+	err = message.Sign(rand.Reader, []byte(""), signer)
+	if err != nil {
+		return err
+	}
+
+	wrap, err := message.MarshalCBOR()
+	if err != nil {
+		return fmt.Errorf("unable to MarshalCBOR %d claimType = %w", cl, err)
+	}
+	if e.message == nil {
+		ccaToken := CcaToken{}
+		e.message = &ccaToken
+	}
+
+	if cl == PlatformClaim {
+		e.message.CcaPlatformToken = &wrap
+	} else {
+		e.message.CcaRealmToken = &wrap
+	}
+	return nil
+}
+
+// Semantic validation, of CCA token
+func (e CcaEvidence) ValidateToken() error {
+
+	// At this point we just check that the both the
+	// Platform and Realm tokens are set correctly.
+	if e.message.CcaPlatformToken == nil {
+		return fmt.Errorf("cca platform token not set")
+	}
+
+	if e.message.CcaRealmToken == nil {
+		return fmt.Errorf("cca realm token not set")
+	}
+
+	return nil
+}
+
+// FromCBOR extracts the message to get the Realm and PlatforToken CBOR bytes
+// and then set the Platform and Realm claims after checking validity
+
+func (e *CcaEvidence) FromCBOR(buf []byte) error {
+	ccaToken := CcaToken{}
+	e.message = &ccaToken
+
+	err := dm.Unmarshal(buf, e.message)
+	if err != nil {
+		return fmt.Errorf("cbor decoding of CCA evidence failed: %w", err)
+	}
+
+	err = e.ValidateToken()
+	if err != nil {
+		return fmt.Errorf("validation of CCA evidence failed: %w", err)
+	}
+
+	// This will set the byte array for Cca Platform Token
+	if err = e.decodeClaims(PlatformClaim); err != nil {
+		return fmt.Errorf("unable to decode platform claims %w", err)
+	}
+
+	// This will set the byte array for Cca Realm Token
+	if err = e.decodeClaims(RealmClaim); err != nil {
+		return fmt.Errorf("unable to decode platform claims %w", err)
+	}
+	return nil
+}
+
+func (e *CcaEvidence) decodeClaims(cl claimType) error {
```

Comment:
this has a similar problem as the signClaim above: it needs not to be a method on the object, it can just be a simpler standalone function